File: examples/language-impl-comparison/elixir/json_extended.ex
defmodule JsonExtended do
  @moduledoc """
  JSON拡張版：コメント・トレーリングカンマ対応。

  標準JSONからの拡張点：
  1. コメント対応（`//` 行コメント、`/* */` ブロックコメント）
  2. トレーリングカンマ許可（配列・オブジェクトの最後の要素の後）
  3. より詳細なエラーメッセージ

  実用的な設定ファイル形式として：
  - `package.json` 風の設定ファイル
  - `.babelrc`, `.eslintrc` など開発ツールの設定
  - VS Code の `settings.json`
  """

  @type json_value ::
          :null
          | boolean()
          | number()
          | String.t()
          | [json_value()]
          | %{String.t() => json_value()}

  @type parse_error :: {:error, String.t()}

  # パース状態
  defmodule State do
    defstruct [:input, :pos]
  end

  @doc """
  JSON拡張文字列をパース。
  """
  @spec parse(String.t()) :: {:ok, json_value()} | parse_error()
  def parse(input) do
    state = %State{input: input, pos: 0}

    with {:ok, state} <- skip_whitespace_and_comments(state),
         {:ok, value, state} <- parse_value(state),
         {:ok, _state} <- skip_whitespace_and_comments(state),
         true <- state.pos >= String.length(state.input) do
      {:ok, value}
    else
      {:error, _} = err -> err
      false -> {:error, "入力の終端に到達していません"}
    end
  end

  # 空白とコメントをスキップ
  defp skip_whitespace_and_comments(state) do
    case skip_ws(state) do
      %State{pos: pos, input: input} = new_state ->
        cond do
          pos >= String.length(input) ->
            {:ok, new_state}

          String.slice(input, pos, 2) == "//" ->
            skip_whitespace_and_comments(skip_line_comment(new_state))

          String.slice(input, pos, 2) == "/*" ->
            case skip_block_comment(new_state) do
              {:ok, state_after} -> skip_whitespace_and_comments(state_after)
              err -> err
            end

          true ->
            {:ok, new_state}
        end
    end
  end

  # 空白文字をスキップ
  defp skip_ws(%State{input: input, pos: pos} = state) do
    len = String.length(input)

    new_pos =
      Enum.reduce_while(pos..(len - 1), pos, fn i, _acc ->
        ch = String.at(input, i)

        if ch in [" ", "\n", "\t", "\r"] do
          {:cont, i + 1}
        else
          {:halt, i}
        end
      end)

    %State{state | pos: new_pos}
  end

  # 行コメントをスキップ
  defp skip_line_comment(%State{input: input, pos: pos} = state) do
    # "//" をスキップ
    new_pos = pos + 2
    len = String.length(input)

    # 改行まで進む
    final_pos =
      Enum.reduce_while(new_pos..(len - 1), new_pos, fn i, _acc ->
        if String.at(input, i) == "\n" do
          {:halt, i + 1}
        else
          {:cont, i + 1}
        end
      end)

    %State{state | pos: final_pos}
  end

  # ブロックコメントをスキップ
  defp skip_block_comment(%State{input: input, pos: pos} = state) do
    # "/*" をスキップ
    new_pos = pos + 2
    len = String.length(input)

    # "*/" を探す
    case find_block_end(input, new_pos, len) do
      {:ok, end_pos} -> {:ok, %State{state | pos: end_pos}}
      :error -> {:error, "ブロックコメントが閉じられていません"}
    end
  end

  defp find_block_end(input, pos, len) when pos + 1 < len do
    if String.slice(input, pos, 2) == "*/" do
      {:ok, pos + 2}
    else
      find_block_end(input, pos + 1, len)
    end
  end

  defp find_block_end(_input, _pos, _len), do: :error

  # 値のパース
  defp parse_value(state) do
    with {:ok, state} <- skip_whitespace_and_comments(state) do
      %State{input: input, pos: pos} = state
      len = String.length(input)

      cond do
        pos >= len ->
          {:error, "予期しないEOF"}

        String.slice(input, pos, 4) == "null" ->
          {:ok, :null, %State{state | pos: pos + 4}}

        String.slice(input, pos, 4) == "true" ->
          {:ok, true, %State{state | pos: pos + 4}}

        String.slice(input, pos, 5) == "false" ->
          {:ok, false, %State{state | pos: pos + 5}}

        String.at(input, pos) == "\"" ->
          parse_string(state)

        String.at(input, pos) == "[" ->
          parse_array(state)

        String.at(input, pos) == "{" ->
          parse_object(state)

        String.at(input, pos) in ["-", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"] ->
          parse_number(state)

        true ->
          {:error, "不正な値"}
      end
    end
  end

  # 文字列リテラルのパース
  defp parse_string(%State{input: input, pos: pos} = state) do
    # 開始の " をスキップ
    new_pos = pos + 1
    len = String.length(input)

    case find_string_end(input, new_pos, len, "") do
      {:ok, str, end_pos} ->
        {:ok, str, %State{state | pos: end_pos}}

      :error ->
        {:error, "文字列が閉じられていません"}
    end
  end

  defp find_string_end(input, pos, len, acc) when pos < len do
    ch = String.at(input, pos)

    case ch do
      "\"" -> {:ok, acc, pos + 1}
      "\\" when pos + 1 < len ->
        next_ch = String.at(input, pos + 1)
        escaped = case next_ch do
          "n" -> "\n"
          "t" -> "\t"
          "r" -> "\r"
          "\\" -> "\\"
          "\"" -> "\""
          _ -> next_ch
        end
        find_string_end(input, pos + 2, len, acc <> escaped)
      _ ->
        find_string_end(input, pos + 1, len, acc <> ch)
    end
  end

  defp find_string_end(_input, _pos, _len, _acc), do: :error

  # 数値のパース
  defp parse_number(%State{input: input, pos: pos} = state) do
    len = String.length(input)

    end_pos =
      Enum.reduce_while(pos..(len - 1), pos, fn i, _acc ->
        ch = String.at(input, i)

        if ch in ["-", "+", ".", "e", "E"] or (ch >= "0" and ch <= "9") do
          {:cont, i + 1}
        else
          {:halt, i}
        end
      end)

    num_str = String.slice(input, pos, end_pos - pos)

    case parse_num_string(num_str) do
      {:ok, num} -> {:ok, num, %State{state | pos: end_pos}}
      :error -> {:error, "不正な数値: #{num_str}"}
    end
  end

  defp parse_num_string(str) do
    cond do
      String.contains?(str, ".") or String.contains?(str, "e") or String.contains?(str, "E") ->
        case Float.parse(str) do
          {num, ""} -> {:ok, num}
          _ -> :error
        end

      true ->
        case Integer.parse(str) do
          {num, ""} -> {:ok, num}
          _ -> :error
        end
    end
  end

  # 配列のパース（トレーリングカンマ対応）
  defp parse_array(%State{input: input, pos: pos} = state) do
    # "[" をスキップ
    new_state = %State{state | pos: pos + 1}

    with {:ok, new_state} <- skip_whitespace_and_comments(new_state) do
      # 空配列チェック
      if String.at(input, new_state.pos) == "]" do
        {:ok, [], %State{new_state | pos: new_state.pos + 1}}
      else
        parse_array_elements(new_state, [])
      end
    end
  end

  defp parse_array_elements(state, acc) do
    with {:ok, value, state} <- parse_value(state),
         {:ok, state} <- skip_whitespace_and_comments(state) do
      new_acc = acc ++ [value]
      ch = String.at(state.input, state.pos)

      case ch do
        "," ->
          state = %State{state | pos: state.pos + 1}
          {:ok, state} = skip_whitespace_and_comments(state)

          # トレーリングカンマチェック
          if String.at(state.input, state.pos) == "]" do
            {:ok, new_acc, %State{state | pos: state.pos + 1}}
          else
            parse_array_elements(state, new_acc)
          end

        "]" ->
          {:ok, new_acc, %State{state | pos: state.pos + 1}}

        _ ->
          {:error, "配列要素の後には ',' または ']' が必要です"}
      end
    end
  end

  # オブジェクトのパース（トレーリングカンマ対応）
  defp parse_object(%State{input: input, pos: pos} = state) do
    # "{" をスキップ
    new_state = %State{state | pos: pos + 1}

    with {:ok, new_state} <- skip_whitespace_and_comments(new_state) do
      # 空オブジェクトチェック
      if String.at(input, new_state.pos) == "}" do
        {:ok, %{}, %State{new_state | pos: new_state.pos + 1}}
      else
        parse_object_pairs(new_state, %{})
      end
    end
  end

  defp parse_object_pairs(state, acc) do
    with {:ok, key, state} <- parse_string(state),
         {:ok, state} <- skip_whitespace_and_comments(state),
         {:ok, state} <- expect_char(state, ":"),
         {:ok, state} <- skip_whitespace_and_comments(state),
         {:ok, value, state} <- parse_value(state),
         {:ok, state} <- skip_whitespace_and_comments(state) do
      new_acc = Map.put(acc, key, value)
      ch = String.at(state.input, state.pos)

      case ch do
        "," ->
          state = %State{state | pos: state.pos + 1}
          {:ok, state} = skip_whitespace_and_comments(state)

          # トレーリングカンマチェック
          if String.at(state.input, state.pos) == "}" do
            {:ok, new_acc, %State{state | pos: state.pos + 1}}
          else
            parse_object_pairs(state, new_acc)
          end

        "}" ->
          {:ok, new_acc, %State{state | pos: state.pos + 1}}

        _ ->
          {:error, "オブジェクト要素の後には ',' または '}' が必要です"}
      end
    end
  end

  # 特定の文字を期待
  defp expect_char(%State{input: input, pos: pos} = state, expected) do
    if String.at(input, pos) == expected do
      {:ok, %State{state | pos: pos + 1}}
    else
      {:error, "'#{expected}' が必要です"}
    end
  end

  @doc """
  JSON値を文字列にレンダリング（検証用）。
  """
  @spec render_to_string(json_value(), integer()) :: String.t()
  def render_to_string(value, indent_level \\ 0)

  def render_to_string(:null, _), do: "null"
  def render_to_string(true, _), do: "true"
  def render_to_string(false, _), do: "false"
  def render_to_string(num, _) when is_number(num), do: to_string(num)
  def render_to_string(str, _) when is_binary(str), do: "\"#{str}\""

  def render_to_string(list, indent_level) when is_list(list) do
    if list == [] do
      "[]"
    else
      indent = String.duplicate("  ", indent_level)
      next_indent = String.duplicate("  ", indent_level + 1)

      items =
        list
        |> Enum.map(&"#{next_indent}#{render_to_string(&1, indent_level + 1)}")
        |> Enum.join(",\n")

      "[\n#{items}\n#{indent}]"
    end
  end

  def render_to_string(map, indent_level) when is_map(map) do
    if map == %{} do
      "{}"
    else
      indent = String.duplicate("  ", indent_level)
      next_indent = String.duplicate("  ", indent_level + 1)

      pairs =
        map
        |> Enum.map(fn {key, val} ->
          "#{next_indent}\"#{key}\": #{render_to_string(val, indent_level + 1)}"
        end)
        |> Enum.join(",\n")

      "{\n#{pairs}\n#{indent}}"
    end
  end

  @doc """
  拡張機能のテスト例。
  """
  def test_extended_json do
    test_cases = [
      {"コメント対応", """
      {
        // これは行コメント
        "name": "test",
        /* これは
           ブロックコメント */
        "version": "1.0"
      }
      """},
      {"トレーリングカンマ", """
      {
        "items": [
          1,
          2,
          3,
        ],
        "config": {
          "debug": true,
          "port": 8080,
        }
      }
      """},
      {"複雑な例", """
      {
        // パッケージ情報
        "name": "my-project",
        "version": "0.1.0",

        /* 依存関係 */
        "dependencies": {
          "core": "1.0",
          "utils": "0.5",
        },

        // スクリプト
        "scripts": {
          "build": "reml build",
          "test": "reml test",
        },
      }
      """}
    ]

    Enum.each(test_cases, fn {name, json_str} ->
      IO.puts("--- #{name} ---")

      case parse(json_str) do
        {:ok, value} ->
          IO.puts("パース成功:")
          IO.puts(render_to_string(value, 0))

        {:error, err} ->
          IO.puts("パースエラー: #{err}")
      end

      IO.puts("")
    end)
  end
end

File: examples/language-impl-comparison/haskell/json_extended.hs
{-|
JSON拡張版：コメント・トレーリングカンマ対応。

標準JSONからの拡張点：
1. コメント対応（`//` 行コメント、`/* */` ブロックコメント）
2. トレーリングカンマ許可（配列・オブジェクトの最後の要素の後）
3. より詳細なエラーメッセージ

実用的な設定ファイル形式として：
- `package.json` 風の設定ファイル
- `.babelrc`, `.eslintrc` など開発ツールの設定
- VS Code の `settings.json`
-}
module JsonExtended
  ( JsonValue(..)
  , ParseError(..)
  , parse
  , renderToString
  , testExtendedJson
  ) where

import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map)
import Data.Char (isDigit, isSpace)
import Data.List (isPrefixOf)
import Control.Monad (void)

-- 型定義

data JsonValue
  = JNull
  | JBool Bool
  | JNumber Double
  | JString String
  | JArray [JsonValue]
  | JObject (Map String JsonValue)
  deriving (Show, Eq)

data ParseError
  = UnexpectedEOF
  | InvalidValue String
  | UnclosedString
  | UnclosedBlockComment
  | ExpectedChar Char
  | InvalidNumber String
  deriving (Show, Eq)

data State = State
  { stateInput :: String
  , statePos :: Int
  } deriving (Show, Eq)

-- パース

parse :: String -> Either ParseError JsonValue
parse input =
  let initialState = State input 0
  in do
    st1 <- skipWhitespaceAndComments initialState
    (value, st2) <- parseValue st1
    finalState <- skipWhitespaceAndComments st2
    if statePos finalState >= length (stateInput finalState)
      then Right value
      else Left $ InvalidValue "入力の終端に到達していません"

-- 空白とコメントをスキップ

skipWhitespaceAndComments :: State -> Either ParseError State
skipWhitespaceAndComments state =
  let stateAfterWs = skipWs state
  in if statePos stateAfterWs >= length (stateInput stateAfterWs)
    then Right stateAfterWs
    else
      let remaining = drop (statePos stateAfterWs) (stateInput stateAfterWs)
      in if "//" `isPrefixOf` remaining
        then skipWhitespaceAndComments (skipLineComment stateAfterWs)
        else if "/*" `isPrefixOf` remaining
          then skipBlockComment stateAfterWs >>= skipWhitespaceAndComments
          else Right stateAfterWs

skipWs :: State -> State
skipWs state@(State input pos)
  | pos >= length input = state
  | isSpace (input !! pos) = skipWs (State input (pos + 1))
  | otherwise = state

skipLineComment :: State -> State
skipLineComment (State input pos) =
  let newPos = pos + 2
      remaining = drop newPos input
      idx = case break (== '\n') remaining of
        (_, "") -> length remaining
        (_, _:_) -> length (takeWhile (/= '\n') remaining) + 1
  in State input (newPos + idx)

skipBlockComment :: State -> Either ParseError State
skipBlockComment (State input pos) =
  let newPos = pos + 2
      remaining = drop newPos input
  in case findSubstring "*/" remaining of
    Nothing -> Left UnclosedBlockComment
    Just idx -> Right $ State input (newPos + idx + 2)

findSubstring :: String -> String -> Maybe Int
findSubstring needle haystack = go 0 haystack
  where
    go _ [] = Nothing
    go n str
      | needle `isPrefixOf` str = Just n
      | otherwise = go (n + 1) (tail str)

-- 値のパース

parseValue :: State -> Either ParseError (JsonValue, State)
parseValue state = do
  cleanState <- skipWhitespaceAndComments state
  let remaining = drop (statePos cleanState) (stateInput cleanState)
  if null remaining
    then Left UnexpectedEOF
    else if "null" `isPrefixOf` remaining
      then Right (JNull, cleanState { statePos = statePos cleanState + 4 })
      else if "true" `isPrefixOf` remaining
        then Right (JBool True, cleanState { statePos = statePos cleanState + 4 })
        else if "false" `isPrefixOf` remaining
          then Right (JBool False, cleanState { statePos = statePos cleanState + 5 })
          else if "\"" `isPrefixOf` remaining
            then parseString cleanState
            else if "[" `isPrefixOf` remaining
              then parseArray cleanState
              else if "{" `isPrefixOf` remaining
                then parseObject cleanState
                else parseNumber cleanState

-- 文字列リテラルのパース

parseString :: State -> Either ParseError (JsonValue, State)
parseString (State input pos) =
  case findStringEnd input (pos + 1) "" of
    Left err -> Left err
    Right (str, endPos) -> Right (JString str, State input endPos)

findStringEnd :: String -> Int -> String -> Either ParseError (String, Int)
findStringEnd input pos acc
  | pos >= length input = Left UnclosedString
  | input !! pos == '"' = Right (acc, pos + 1)
  | input !! pos == '\\', pos + 1 < length input =
      let escaped = case input !! (pos + 1) of
            'n' -> "\n"
            't' -> "\t"
            'r' -> "\r"
            '\\' -> "\\"
            '"' -> "\""
            ch -> [ch]
      in findStringEnd input (pos + 2) (acc ++ escaped)
  | otherwise = findStringEnd input (pos + 1) (acc ++ [input !! pos])

-- 数値のパース

parseNumber :: State -> Either ParseError (JsonValue, State)
parseNumber (State input pos) =
  let numStr = takeWhile isNumChar (drop pos input)
      endPos = pos + length numStr
  in case reads numStr :: [(Double, String)] of
    [(num, "")] -> Right (JNumber num, State input endPos)
    _ -> Left $ InvalidNumber numStr

isNumChar :: Char -> Bool
isNumChar ch = ch `elem` "+-." || ch == 'e' || ch == 'E' || isDigit ch

-- 配列のパース（トレーリングカンマ対応）

parseArray :: State -> Either ParseError (JsonValue, State)
parseArray (State input pos) = do
  let stateAfterBracket = State input (pos + 1)
  cleanState <- skipWhitespaceAndComments stateAfterBracket
  let remaining = drop (statePos cleanState) (stateInput cleanState)
  if "]" `isPrefixOf` remaining
    then Right (JArray [], cleanState { statePos = statePos cleanState + 1 })
    else parseArrayElements cleanState []

parseArrayElements :: State -> [JsonValue] -> Either ParseError (JsonValue, State)
parseArrayElements state acc = do
  (value, stateAfterValue) <- parseValue state
  cleanState <- skipWhitespaceAndComments stateAfterValue
  let newAcc = acc ++ [value]
      remaining = drop (statePos cleanState) (stateInput cleanState)
  if null remaining
    then Left UnexpectedEOF
    else case head remaining of
      ',' -> do
        let stateAfterComma = cleanState { statePos = statePos cleanState + 1 }
        stateAfterWs <- skipWhitespaceAndComments stateAfterComma
        let nextRemaining = drop (statePos stateAfterWs) (stateInput stateAfterWs)
        if "]" `isPrefixOf` nextRemaining
          then Right (JArray newAcc, stateAfterWs { statePos = statePos stateAfterWs + 1 })
          else parseArrayElements stateAfterWs newAcc
      ']' -> Right (JArray newAcc, cleanState { statePos = statePos cleanState + 1 })
      _ -> Left $ ExpectedChar ','

-- オブジェクトのパース（トレーリングカンマ対応）

parseObject :: State -> Either ParseError (JsonValue, State)
parseObject (State input pos) = do
  let stateAfterBrace = State input (pos + 1)
  cleanState <- skipWhitespaceAndComments stateAfterBrace
  let remaining = drop (statePos cleanState) (stateInput cleanState)
  if "}" `isPrefixOf` remaining
    then Right (JObject Map.empty, cleanState { statePos = statePos cleanState + 1 })
    else parseObjectPairs cleanState Map.empty

parseObjectPairs :: State -> Map String JsonValue -> Either ParseError (JsonValue, State)
parseObjectPairs state acc = do
  (JString key, stateAfterKey) <- parseString state
  cleanState1 <- skipWhitespaceAndComments stateAfterKey
  let remaining1 = drop (statePos cleanState1) (stateInput cleanState1)
  if not (":" `isPrefixOf` remaining1)
    then Left $ ExpectedChar ':'
    else do
      let stateAfterColon = cleanState1 { statePos = statePos cleanState1 + 1 }
      cleanState2 <- skipWhitespaceAndComments stateAfterColon
      (value, stateAfterValue) <- parseValue cleanState2
      cleanState3 <- skipWhitespaceAndComments stateAfterValue
      let newAcc = Map.insert key value acc
          remaining3 = drop (statePos cleanState3) (stateInput cleanState3)
      if null remaining3
        then Left UnexpectedEOF
        else case head remaining3 of
          ',' -> do
            let stateAfterComma = cleanState3 { statePos = statePos cleanState3 + 1 }
            stateAfterWs <- skipWhitespaceAndComments stateAfterComma
            let nextRemaining = drop (statePos stateAfterWs) (stateInput stateAfterWs)
            if "}" `isPrefixOf` nextRemaining
              then Right (JObject newAcc, stateAfterWs { statePos = statePos stateAfterWs + 1 })
              else parseObjectPairs stateAfterWs newAcc
          '}' -> Right (JObject newAcc, cleanState3 { statePos = statePos cleanState3 + 1 })
          _ -> Left $ ExpectedChar ','

-- レンダリング

renderToString :: JsonValue -> Int -> String
renderToString value indentLevel =
  let indent = replicate (indentLevel * 2) ' '
      nextIndent = replicate ((indentLevel + 1) * 2) ' '
  in case value of
    JNull -> "null"
    JBool True -> "true"
    JBool False -> "false"
    JNumber num -> show num
    JString str -> "\"" ++ str ++ "\""
    JArray items ->
      if null items
        then "[]"
        else
          let itemsStr = unlines $ map (\item -> nextIndent ++ renderToString item (indentLevel + 1) ++ ",") (init items)
                      ++ [nextIndent ++ renderToString (last items) (indentLevel + 1)]
          in "[\n" ++ itemsStr ++ "\n" ++ indent ++ "]"
    JObject pairs ->
      if Map.null pairs
        then "{}"
        else
          let pairsList = Map.toList pairs
              pairsStr = unlines $ map (\(key, val) -> nextIndent ++ "\"" ++ key ++ "\": " ++ renderToString val (indentLevel + 1) ++ ",") (init pairsList)
                        ++ [nextIndent ++ "\"" ++ fst (last pairsList) ++ "\": " ++ renderToString (snd (last pairsList)) (indentLevel + 1)]
          in "{\n" ++ pairsStr ++ "\n" ++ indent ++ "}"

-- テスト

testExtendedJson :: IO ()
testExtendedJson = do
  let testCases =
        [ ("コメント対応", "{\n  // これは行コメント\n  \"name\": \"test\",\n  /* これは\n     ブロックコメント */\n  \"version\": \"1.0\"\n}\n")
        , ("トレーリングカンマ", "{\n  \"items\": [\n    1,\n    2,\n    3,\n  ],\n  \"config\": {\n    \"debug\": true,\n    \"port\": 8080,\n  }\n}\n")
        ]
  mapM_ (\(name, jsonStr) -> do
    putStrLn $ "--- " ++ name ++ " ---"
    case parse jsonStr of
      Right value -> do
        putStrLn "パース成功:"
        putStrLn $ renderToString value 0
      Left err ->
        putStrLn $ "パースエラー: " ++ show err
    putStrLn ""
    ) testCases